### PR TITLE
Implementation of the Hessian application to accelerate convergence for load inversion

### PR DESCRIPTION
**Pull request by mguddati**
Implementation of the Hessian application to accelerate convergence finalized for load inversion. May need some modifications for parallel and for material inversion
see https://github.com/idaholab/isopod/issues/6

### CODE REVIEW

File: src/executioners/OptimizeSolve.C [hunk 1]
```diff
@@ -289,12 +318,12 @@ OptimizeSolve::objectiveFunction()
 {
   _form_function->updateParameters(*_parameters.get());
 
-  _problem.execute(EXEC_FORWARD);
+//  _problem.execute(EXEC_FORWARD);
```
Comment (mguddati) [inline src/executioners/OptimizeSolve.C:321]: @lynnmunday @zachmprince  Do we need this? The tests pass without this. If this is needed at times, this may need to be included in the applyHessian routine.

File: src/executioners/OptimizeSolve.C [hunk 2]
```diff
@@ -289,12 +318,12 @@ OptimizeSolve::objectiveFunction()
...
   bool multiapp_passed = true;
   if (!_problem.execMultiApps(EXEC_FORWARD))
     multiapp_passed = false;
-  if (_solve_on.contains(EXEC_FORWARD))
-    _inner_solve->solve();
+//  if (_solve_on.contains(EXEC_FORWARD))
```
Comment (mguddati) [inline src/executioners/OptimizeSolve.C:325]: @lynnmunday @zachmprince  Do we need this? The tests pass without this. If this is needed at times, this may need to be included in the applyHessian routine.

Comment (mguddati) [reply src/executioners/OptimizeSolve.C:325]: ... there are similar things elsewhere (innersolve).

File: src/executioners/OptimizeSolve.C [hunk 3]
```diff
@@ -263,16 +261,47 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
 }
 
 PetscErrorCode
-OptimizeSolve::hessianFunctionWrapper(Tao /*tao*/, Vec x, Mat hessian, Mat /*pc*/, void * ctx)
+OptimizeSolve::hessianFunctionWrapper(Tao /*tao*/, Vec x, Mat /*hessian*/, Mat /*pc*/, void * ctx)
 {
   auto * solver = static_cast<OptimizeSolve *>(ctx);
```
Comment (lynnmunday) [inline src/executioners/OptimizeSolve.C:266]: I think this entire function can be empty.  I don't think that it is doing anything that hasn't already been done by the first MatShellSetOperation on ln 130-134.  This wrapper is setting x into the solvers _parameters but those _parameters are never used in the hessian wrappers or functions.  Should these _parameters be used in applyHessianWrapper?

File: src/executioners/OptimizeSolve.C [hunk 4]
```diff
@@ -263,16 +261,47 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
...
+  libMesh::PetscVector<Number> sbar(s, solver->_my_comm);
+  libMesh::PetscVector<Number> Hsbar(Hs, solver->_my_comm);
+  return solver->applyHessian(sbar,Hsbar);
+}
 
-  solver->hessianFunction(mat);
+PetscErrorCode
+OptimizeSolve::applyHessian(libMesh::PetscVector<Number> &s, libMesh::PetscVector<Number> &Hs)
+{
+  _form_function->updateParameters(s);
```
Comment (lynnmunday) [inline src/executioners/OptimizeSolve.C:290]: This is where the _parameters are gotten from the solver with the other wrappers using:
```_form_function->updateParameters(*_parameters.get());```

File: src/executioners/OptimizeSolve.C [hunk 5]
```diff
@@ -263,16 +261,47 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
...
+PetscErrorCode
+OptimizeSolve::applyHessian(libMesh::PetscVector<Number> &s, libMesh::PetscVector<Number> &Hs)
+{
+  _form_function->updateParameters(s); 
+      // What happens for material inversion when the Hessian 
+      // is dependent on the parameters? Deal with it later???
+  if (!_problem.execMultiApps(EXEC_FORWARD))
+    mooseError("Forward solve multiapp failed!");
+// delete this?
+//  if (_solve_on.contains(EXEC_FORWARD))
```
Comment (lynnmunday) [inline src/executioners/OptimizeSolve.C:296]: we can delete all these commented out lines and the inner_solve->solve()

File: src/optimizationreporters/OptimizationReporter.C [hunk 6]
```diff
@@ -121,3 +122,10 @@ OptimizationReporter::computeObjective()
 
   return val;
 }
+
+void
+OptimizationReporter::setMisfitToSimulatedValues()
```
Comment (lynnmunday) [inline src/optimizationreporters/OptimizationReporter.C:127]: This seems weird.  Maybe the variable is named wrong and should be adjoint parameters instead of misfit parameters.  Then we aren't setting misfit_values to something other than the misfit

File: test/tests/executioners/tests [hunk 7]
```diff
@@ -19,22 +19,21 @@
...
-    csvdiff = qm_multi_param_out_OptimizationReporter_0001.csv
-    requirement = "Solve with multiple parameters."
-  []
+
+#  [quadratic_taontr]
+#    type = CSVDiff
+#    cli_args = "Executioner/tao_solver=TAONTR"
+#    " Executioner/petsc_options_iname='-tao_ntr_min_radius -tao_ntr_max_radius -tao_ntr_init_type'"
+#    " Executioner/petsc_options_value='0 1e16 constant'"
+#    input = quadratic_minimize.i
```
Comment (lynnmunday) [inline test/tests/executioners/tests:28]: Can these be fixed?

File: test/tests/optimizationreporter/objective_gradient_minimize/mechanics/master.i [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  tao_solver = taonls
+  petsc_options_iname = '-tao_gttol -tao_max_it -tao_nls_pc_type -tao_nls_ksp_type'
+  petsc_options_value = '1e-5 50 none cg'
+  verbose = true
+
+#  type = Optimize
+#  tao_solver = taolmvm
+#  petsc_options_iname = '-tao_gatol -tao_grtol'
+#  petsc_options_value = '1e-6 1e-6'
+#  verbose = true
```
Comment (lynnmunday) [inline test/tests/optimizationreporter/objective_gradient_minimize/mechanics/master.i:35]: Is there a -tao_test_hessian test?

**Review by lynnmunday — COMMENTED** [review]
This works for the problems you tried here but I couldn't get the Hessian to work with
```isopod/test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction```

do you know what the problem might be.  What do you have to look for when something isn't working with the hessian?  I don't know how to debug it.  It just keeps iterating.    I made most of the below changes and I'll push it to this PR for you to look at.

**Comment by lynnmunday** [comment]
I can't figure out how to push to your branch.  I'm not sure what branch on idaholabs we should be pushing to.  I'll need help syncing my git repo with teh repo on idaholabs.

**Comment by zachmprince** [comment]
@mguddati I just want to make sure I understand how this works. Basically, you implemented a matrix-free Hessian to be able to use quadratic optimization algorithms. From what I understand a matrix-free Hessian is the action of the Hessian on the parameter vector, which I believe is equivalent to the gradient? How the optimization algorithms work is by inverting the Hessian by using some sort of matrix-free method (probably something like GMRes). So the applyHessian is called multiple times during the inversion process, right? Which means there are multiple, if not many, solves during a single optimization step?

**Comment by mguddati** [comment]
> 

@zachmprince - sorry for the (very) late reply. You are absolutely right in all aspects (except, I don't understand when you say Hessian*incremental parameter vector is like the gradient - it is only true in that they live in the same space, but they are not quite the same - perhaps this is what you meant).

File: src/executioners/OptimizeSolve.C (hunk 3 above)
Comment (zachmprince) [reply src/executioners/OptimizeSolve.C:269]: Yes, let's empty out this function. Eventually we can assemble an actual Hessian as a preconditioner here.

File: src/executioners/OptimizeSolve.C [hunk 9]
```diff
@@ -263,16 +259,58 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
...
+  void * ctx;
+  MatShellGetContext(H, &ctx);
+  auto * solver = static_cast<OptimizeSolve *>(ctx);
+  libMesh::PetscVector<Number> sbar(s, solver->_my_comm);
+  libMesh::PetscVector<Number> Hsbar(Hs, solver->_my_comm);
+  return solver->applyHessian(sbar, Hsbar);
+}
+
+PetscErrorCode
+OptimizeSolve::applyHessian(libMesh::PetscVector<Number> & s, libMesh::PetscVector<Number> & Hs)
```
Comment (zachmprince) [inline src/executioners/OptimizeSolve.C:293]: Could you move this function below `gradientFunction`? Feel like it makes more sense to put there.

File: src/executioners/OptimizeSolve.C [hunk 10]
```diff
@@ -263,16 +259,58 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
...
+  if (!_problem.execMultiApps(EXEC_HOMOGENEOUS_FORWARD))
+    mooseError("Forward solve multiapp failed!");
+  _obj_iterate++;
+
+  _form_function->setMisfitToSimulatedValues();
+  if (!_problem.execMultiApps(EXEC_ADJOINT))
+    mooseError("Adjoint solve multiapp failed!");
+  _grad_iterate++;
+  _form_function->computeGradient(Hs);
+  _hess_iterate++;
```
Comment (zachmprince) [inline src/executioners/OptimizeSolve.C:313]: Not sure this increment is appropriate here.

File: src/executioners/OptimizeSolve.C [hunk 11]
```diff
@@ -289,13 +327,9 @@ Real
 OptimizeSolve::objectiveFunction()
 {
   _form_function->updateParameters(*_parameters.get());
-
-  _problem.execute(EXEC_FORWARD);
```
Comment (zachmprince) [inline src/executioners/OptimizeSolve.C:293]: Please don't remove this.

File: src/executioners/OptimizeSolve.C [hunk 12]
```diff
@@ -289,13 +327,9 @@ Real
...
   bool multiapp_passed = true;
   if (!_problem.execMultiApps(EXEC_FORWARD))
     multiapp_passed = false;
-  if (_solve_on.contains(EXEC_FORWARD))
-    _inner_solve->solve();
```
Comment (zachmprince) [inline src/executioners/OptimizeSolve.C:298]: Please don't remove this.

File: src/executioners/OptimizeSolve.C [hunk 13]
```diff
@@ -263,16 +259,58 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
...
-  solver->hessianFunction(mat);
+  // What happens for material inversion when the Hessian
+  // is dependent on the parameters? Deal with it later???
+  // see notes on how this needs to change for Material inversion
+  _form_function->updateParameters(s);
+  if (!_problem.execMultiApps(EXEC_HOMOGENEOUS_FORWARD))
+    mooseError("Forward solve multiapp failed!");
+  _obj_iterate++;
+
+  _form_function->setMisfitToSimulatedValues();
```
Comment (zachmprince) [inline src/executioners/OptimizeSolve.C:308]: Can we think of a more general name for this? Maybe something like `computeHomogeneousObjective`?

File: test/tests/executioners/quadratic_minimize.i [hunk 14]
```diff
@@ -6,7 +6,7 @@
   parameter_names = 'results'
   num_values = 3
   initial_condition = '5 8 1'
-  objective = 1.0
+  objective = 0.0
```
Comment (zachmprince) [inline test/tests/executioners/quadratic_minimize.i:9]: Why this change?

File: test/tests/executioners/tests [hunk 15]
```diff
@@ -21,20 +21,20 @@
   []
   [quadratic_taontr]
     type = CSVDiff
-    max_threads = 1
     cli_args = "Executioner/tao_solver=TAONTR"
     " Executioner/petsc_options_iname='-tao_ntr_min_radius -tao_ntr_max_radius -tao_ntr_init_type'"
     " Executioner/petsc_options_value='0 1e16 constant'"
     input = quadratic_minimize.i
     csvdiff = quadratic_minimize_out_OptimizationReporter_0001.csv
+    skip = "Hessian broken"
```
Comment (zachmprince) [inline test/tests/executioners/tests:29]: I think the solver needs to work with this example. It is absolute simplest optimization problem.

File: test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/master.i [hunk 16]
```diff
@@ -22,21 +22,35 @@
...
+                        0.8 1.8 0'
   measurement_values = '207 204 185 125'
 []
 
 [Executioner]
   type = Optimize
-  tao_solver = taolmvm
-  petsc_options_iname = '-tao_gatol'#
-  petsc_options_value = '1e-4'
+  # tao_solver = taobncg
```
Comment (zachmprince) [inline test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/master.i:33]: Remove all commented lines.

File: test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/plot_iterations.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels
+plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels
+plt.rc('legend', fontsize=7)    # legend fontsize
+plt.rc('figure', titlesize=SMALL_SIZE)  # fontsize of the figure title
+#-------------------------------------------------------------------------------
+df_nls = pd.read_csv('nls_optInfo_0001.csv')
+df_cg = pd.read_csv('cg_optInfo_0001.csv')
+df_nm = pd.read_csv('nm_optInfo_0001.csv')
+df_lmvm = pd.read_csv('lmvm_optInfo_0001.csv')
+print('fracture dataframe description: ',df_nls.columns.values)
```
Comment (zachmprince) [inline test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/plot_iterations.py:33]: If we are going to add a python utility it needs to be general, like using input arguments instead of hardcoded filenames. Also, using json format with ReporterReader would be way cleaner.

**Review by zachmprince — CHANGES_REQUESTED** [review]

File: src/executioners/OptimizeSolve.C [hunk 18]
```diff
@@ -263,16 +259,58 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
...
+  //    mooseError("Hessian based optimization algorithms require a sub-app with:\n"
+  //               "   execute_on = HOMOGENEOUS_FORWARD");
 
-  solver->hessianFunction(mat);
+  // What happens for material inversion when the Hessian
+  // is dependent on the parameters? Deal with it later???
+  // see notes on how this needs to change for Material inversion
+  _form_function->updateParameters(s);
+  if (!_problem.execMultiApps(EXEC_HOMOGENEOUS_FORWARD))
+    mooseError("Forward solve multiapp failed!");
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply src/executioners/OptimizeSolve.C:305]: I removed these errors because the test objects will fail because they don't have a subapp.  Maybe we need to do something about this interface in OptimizeSolve.C. Like can we derive something from it for when we have subapps.... I don't know if that will work since tao needs to be able to call static functions, I don't think those can be polymorphic.  Or we need to redesign the test objects to test the code in the way we are using it. I would like the test objects to provide a simple example of using tao directly.

File: test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/plot_iterations.py (hunk 17 above)
Comment (lynnmunday) [reply test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/plot_iterations.py:33]: @mguddati you can delete this python file.  I added it but it doesn't belong in the test directory.  I'm fine with json output, I haven't tried it but it should work

**Comment by lynnmunday** [comment]
@zachmprince how do you trigger the tests to run.  I think I lost permission to do that.  WIll you trigger them

File: src/executioners/OptimizeSolve.C (hunk 3 above)
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:269]: Kept the function and moved MatShellSetOperation into this routine.

File: src/executioners/OptimizeSolve.C (hunk 9 above)
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:293]: Done.

File: src/executioners/OptimizeSolve.C [hunk 19]
```diff
@@ -263,16 +259,58 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
...
+}
+
+PetscErrorCode
+OptimizeSolve::applyHessian(libMesh::PetscVector<Number> & s, libMesh::PetscVector<Number> & Hs)
+{
 
-  libMesh::PetscMatrix<Number> mat(hessian, solver->_my_comm);
+  //  if (!_problem.hasMultiApps(EXEC_HOMOGENEOUS_FORWARD))
+  //    mooseError("Hessian based optimization algorithms require a sub-app with:\n"
+  //               "   execute_on = HOMOGENEOUS_FORWARD");
```
[the remark this answers is not in the document]
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:298]: Uncommented this.

File: src/executioners/OptimizeSolve.C (hunk 18 above)
[the remark this answers is not in the document]
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:305]: Done.

File: src/executioners/OptimizeSolve.C (hunk 13 above)
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:308]: Let's leave it as is as it is precise and descriptive (setting values on the reporter). We will revisit later.

File: src/executioners/OptimizeSolve.C [hunk 20]
```diff
@@ -263,16 +259,58 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
...
+  // is dependent on the parameters? Deal with it later???
+  // see notes on how this needs to change for Material inversion
+  _form_function->updateParameters(s);
+  if (!_problem.execMultiApps(EXEC_HOMOGENEOUS_FORWARD))
+    mooseError("Forward solve multiapp failed!");
+  _obj_iterate++;
+
+  _form_function->setMisfitToSimulatedValues();
+  if (!_problem.execMultiApps(EXEC_ADJOINT))
+    mooseError("Adjoint solve multiapp failed!");
```
[the remark this answers is not in the document]
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:310]: Done.

File: src/executioners/OptimizeSolve.C (hunk 10 above)
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:313]: Done - removed _obj_iterate++ and _grad_iterate++

File: src/executioners/OptimizeSolve.C (hunk 11 above)
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:293]: done.

File: src/executioners/OptimizeSolve.C (hunk 12 above)
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:298]: done.

File: src/executioners/OptimizeSolve.C [hunk 21]
```diff
@@ -305,33 +339,13 @@ void
 OptimizeSolve::gradientFunction(libMesh::PetscVector<Number> & gradient)
 {
   _form_function->updateParameters(*_parameters.get());
-
-  _problem.execute(EXEC_ADJOINT);
```
[the remark this answers is not in the document]
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:309]: done

File: src/executioners/OptimizeSolve.C [hunk 22]
```diff
@@ -305,33 +339,13 @@ void
...
   if (!_problem.execMultiApps(EXEC_ADJOINT))
     mooseError("Adjoint solve multiapp failed!");
-  if (_solve_on.contains(EXEC_ADJOINT))
-    _inner_solve->solve();
```
[the remark this answers is not in the document]
Comment (mguddati) [reply src/executioners/OptimizeSolve.C:313]: done

File: test/tests/executioners/quadratic_minimize.i (hunk 14 above)
Comment (mguddati) [reply test/tests/executioners/quadratic_minimize.i:9]: these tests are no longer relevant. We need to put in unit tests where we don't build the hessian but apply it.

File: test/tests/executioners/tests (hunk 15 above)
Comment (mguddati) [reply test/tests/executioners/tests:29]: these tests are no longer relevant. We need to put in unit tests where we don't build the hessian but apply it.

File: test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/master.i (hunk 16 above)
Comment (mguddati) [reply test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/master.i:33]: We will do that as a separate commit later.

File: test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/plot_iterations.py (hunk 17 above)
Comment (mguddati) [reply test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/plot_iterations.py:33]: done

File: test/tests/optimizationreporter/objective_gradient_minimize/mechanics/master.i (hunk 8 above)
Comment (mguddati) [reply test/tests/optimizationreporter/objective_gradient_minimize/mechanics/master.i:35]: discussed. not relevant.

**Comment by mguddati** [comment]
@zachmprince - we addressed your comments. Please review.

File: test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/master.i (hunk 16 above)
Comment (lynnmunday) [reply test/tests/optimizationreporter/objective_gradient_minimize/bc_load_linearFunction/master.i:33]: I've been commenting out solver settings in a lot of tests so that I can easily switch between the different optimization parameters.  I'll add an issue to delete these.

File: test/tests/executioners/tests (hunk 15 above)
Comment (lynnmunday) [reply test/tests/executioners/tests:29]: New unit tests will be hard to make because the apply hessian is done in optimize solve and it doesn't get it's data from a reporter.  We were able to test gradient based optimization because that only required us to make a new reporter that would give us the gradient value.  Will we need to derive off OptimizeSolve in order to create a new apply hessian method?  I don't know if that makes sense.

File: test/tests/executioners/quadratic_minimize.i (hunk 14 above)
Comment (lynnmunday) [reply test/tests/executioners/quadratic_minimize.i:9]: I"ll add an issue for this.

**Comment by lynnmunday** [comment]
@zachmprince Will you review, I can come over and explain the changes when you're in the office.  I've created several issues to cover some of the requests you wanted in the previous review that were not made in the current one.

**Review by zachmprince — APPROVED** [review]
